### ISSUE TITLE
package:dwds is writing null into the json values for some objects

### ISSUE THREAD

**Issue by devoncarew** [dart-lang/webdev#575]
I'm reading values out of the json objects returned by package:dwds as part of debugging a Flutter for web app. I see `JsonNull` values in strange places, like where I'd expect to see `ObjRef.id`'s:

```
java.lang.UnsupportedOperationException: JsonNull
	at com.google.gson.JsonElement.getAsString(JsonElement.java:192)
	at org.dartlang.vm.service.element.ObjRef.getId(ObjRef.java:44)
	at io.flutter.inspector.EvalOnDartLibrary.lambda$getObjHelper$6(EvalOnDartLibrary.java:171)
	at io.flutter.inspector.EvalOnDartLibrary.lambda$addRequest$1(EvalOnDartLibrary.java:79)
```

I don't have a repro currently, but have see it twice, in two different places.

This is from the latest from Flutter at head.

**Comment by devoncarew** [comment]
Hmm, the `JsonNull` value is an artifact of the Java json library we're using. The issue however is that the `ObjRef`'s `id` is null. Here's the returned json:

> {"type":"@Instance","id":null,"kind":"String","class":{"type":"@Class","id":"dart:core:string","name":"String"},"valueAsString":"{\"description\":\"[root]\",\"type\":\"_ElementDiagnosticableTreeNode\",\"style\":\"dense\",\"hasChildren\":true,\"allowWrap\":false,\"objectId\":\"inspector-0\",\"valueId\":\"inspector-1\",\"summaryTree\":true,\"locationId\":0,\"creationLocation\":{\"file\":\"org-dartlang-app:///packages/flutter/src/widgets/binding.dart\",\"line\":791,\"column\":26,\"parameterLocations\":[{\"file\":null,\"line\":792,\"column\":7,\"name\":\"container\"},{\"file\":null,\"line\":793,\"column\":7,\"name\":\"debugShortDescription\"},{\"file\":null,\"line\":794,\"column\":7,\"name\":\"child\"}]},\"children\":[],\"widgetRuntimeType\":\"RenderObjectToWidgetAdapter<RenderBox>\",\"stateful\":false}","valueAsStringIsTruncated":false}

**Comment by grouma** [comment]
Hmm.. Is `remoteObject.objectId` null here:
https://github.com/dart-lang/webdev/blob/d8f7a4b7f5e6c4443aeeaf1e329072dd9f6e7fa6/dwds/lib/src/debugging/instance.dart#L78

**Comment by devoncarew** [comment]
In terms of the id for the object that we pass in? I don't know. I can try and diagnose when I see it next.

**Comment by grouma** [comment]
This is the only place `instanceFor` is called and we use the `objectId` provided to us:
https://github.com/dart-lang/webdev/blob/master/dwds/lib/src/debugging/inspector.dart#L278

**Comment by alan-knight** [comment]
I think that's an InstanceRef, not an Instance, so it could be coming back from quite a few more places. I'd be suspicious that we might have a reference to something we can't properly materialize as a Dart object and send back something malformed.

**Comment by devoncarew** [comment]
I'll see where we're getting the `objectId` that we're passing in next time I see this.

**Comment by alan-knight** [comment]
Actually, I think we'd expect to always see this for strings. This isn't the matching version, but for strings we call _primitiveInstance, and we don't give any of those an objectId. That should be easy to fix.

https://github.com/dart-lang/webdev/blob/b163bda0d28b18a01065332fdff5c4a4b63adbbb/dwds/lib/src/debugging/instance.dart#L92
and then 
https://github.com/dart-lang/webdev/blob/b163bda0d28b18a01065332fdff5c4a4b63adbbb/dwds/lib/src/debugging/instance.dart#L40

**Comment by devoncarew** [comment]
OK.

For strings when debugging against the VM, when we need a string value we check to see whether the 'truncated' value is set. If it is, then we cann getObject with the object id, and the char range, in order to retrive the full string value.

I don't know if that's whats happening here; we'e seeing this issue from some Inspector calls from IntelliJ, when communicating with a flutter web client.

**Comment by alan-knight** [comment]
Hmm. That would make sense - it might only show up for strings above a certain length, and that's why we haven't hit it immediately. If I push a proposed fix to a branch, can you test against that easily?

**Comment by devoncarew** [comment]
Yes, I think so - I could manually edit my .packages file.

**Comment by devoncarew** [comment]
OK, the inspector is trying to retrieve the full contents for a String object.

We have an InstanceRef, which is a String type, and we're calling getObject() on it, and expecting an Instance back.

Here is the contents of the InstanceRef; the `id` is `null`.

> {"type":"@Instance","id":null,"kind":"String","class":{"type":"@Class","id":"dart:core:string","name":"String"},"valueAsString":"{\"description\":\"[root]\",\"type\":\"_ElementDiagnosticableTreeNode\",\"style\":\"dense\",\"hasChildren\":true,\"allowWrap\":false,\"objectId\":\"inspector-0\",\"valueId\":\"inspector-1\",\"summaryTree\":true,\"locationId\":0,\"creationLocation\":{\"file\":\"org-dartlang-app:///packages/flutter/src/widgets/binding.dart\",\"line\":791,\"column\":26,\"parameterLocations\":[{\"file\":null,\"line\":792,\"column\":7,\"name\":\"container\"},{\"file\":null,\"line\":793,\"column\":7,\"name\":\"debugShortDescription\"},{\"file\":null,\"line\":794,\"column\":7,\"name\":\"child\"}]},\"children\":[],\"widgetRuntimeType\":\"RenderObjectToWidgetAdapter<RenderBox>\",\"stateful\":false}","valueAsStringIsTruncated":false}

**Comment by devoncarew** [comment]
I do have a workaround for this now in IntelliJ (but I think we'll still want to support `getObject()` for String objects).

**Comment by alan-knight** [comment]
This is a little trickier than I initially thought. Chrome isn't giving us an objectId for strings. And it doesn't appear to be related to truncation, because we aren't truncating anywhere, and that's false in the JSON you provided. So it seems like either we would have to maintain our own registry of object values that have been returned, or we could do some kind of hack where we use the String's value as its own objectId, with some kind of prefix that should lead to it not being confuse-able with a regular Chrome objectId. Obviously, I chose the second :-) It's https://github.com/dart-lang/webdev/pull/581 if anyone has opinions as to its desirability and if it actually fixes the problem.

**Comment by alan-knight** [comment]
Fixed by #581
